integration/plugin/authz: port tests from integration-cli

dsheets opened the pull request:
**- What I did**

I ported the authz plugin tests from `integration-cli` using the CLI to `integration/plugin/authz` using the API.

At least some of this work is required for #33658 and #33375.

**- How I did it**

I mechanically changed the structure of the tests to fit the new API test style, replaced CLI use with API helper utility function calls, and then fixed up build and test failures.

**- How to verify it**

The tests execute and pass. Side-by-side visual comparison can demonstrate that they test the functionally equivalent behaviors.

**- Description for the changelog**

N/A

**- A picture of a cute animal (not mandatory but encouraged)**

![long-tailed planigale](https://pbs.twimg.com/media/BlPf4IMIYAAGm8w.jpg:large)

[*Planigale ingrami*](https://en.wikipedia.org/wiki/Long-tailed_planigale), or the long-tailed planigale, is believed to be the smallest marsupial on Earth. It is native to Australia's Top End where it is abundant but rarely seen. The average weight for males is 4.2g and for females is 4.3g.

/cc @dnephin @thaJeztah

Commit:
integration/plugin/authz: port tests from integration-cli

Signed-off-by: David Sheets <dsheets@docker.com>

thaJeztah commented:
Thanks! Looks like there's some linting failures;

```
14:07:06 integration/internal/api/image/image.go:12:1:warning: exported function Save should have comment or be unexported (golint)
14:07:06 integration/internal/api/image/image.go:25:2:warning: 'if err != nil { return err }; return nil' can be simplified to 'return err' (S1013) (gosimple)
14:07:06 integration/plugin/authz/authz_plugin_test.go:344:13:warning: ineffectual assignment to err (ineffassign)
14:07:06 integration/util/requirement/requirement.go:14:29:warning: unnecessary conversion (unconvert)

dsheets commented:
Yes, I've just fixed those. For some reason, `make validate` hangs on my machine recently. It would be really great to have some documentation of exactly what tests are run by CI and how. I've had `make validate` pass in the past but CI's linting fail. I also have tested this PR successfully locally with something like `DOCKER_GRAPHDRIVER=vfs TESTFLAGS='-test.run TestAuthZPlugin*' make test-integration`.

dnephin commented:
#34911 should make it a lot more obvious about what is run by CI. Needs another LGTM to merge

> I've had make validate pass in the past but CI's linting fail.

If you encounter this again please let me know

dnephin commented on integration/util/requirement/requirement.go:
```diff
@@ -0,0 +1,29 @@
+
+	resp, err := client.Get(url)
+	if err != nil && strings.Contains(err.Error(), "use of closed network connection") {
+		panic(fmt.Sprintf("Timeout for GET request on %s", url))
```
It seems strange to panic here, especially on a timeout.

How about add a `t LogT` argument to this function and do this instead:

```go
t.Logf("Timeout ...")
return false
```

dsheets replied:
This comes directly from <https://github.com/moby/moby/blob/master/integration-cli/requirements_test.go#L106> which I did not write. I don't think tests that require hub connectivity should allow test runs to pass by skipping them and emitting a logging message that will be missed/ignored in the thousands of lines of spew.

dnephin replied:
We should fix these problem instead of carrying them over into the new test suite.

> I don't think tests that require hub connectivity should allow test runs to pass by skipping them and emitting a logging message that will be missed/ignored in the thousands of lines of spew.

If you think this should be a fatal error you can use `t.Fatal()`, but the purpose of these helpers is to detect state and skip tests that can't be run, so I don't think that's really appropriate. Skipping the test is the right behaviour. There is no reason to panic here.

dnephin commented on integration/testdata/main_test.go:
```diff
@@ -0,0 +1 @@
+package testdata
```
I don't think testdata should be a package. If this is because hack/make/test-integration tries to run this as a test package how about we just add testdata to the exclude list?

dnephin commented on integration/testdata/https/main_test.go:
```diff
@@ -0,0 +1 @@
+package https
```
also doesn't need to be a package

dnephin commented on integration/plugin/authz/main_test.go:
```diff
@@ -0,0 +1,177 @@
+	return func() {
+		if d != nil {
+			d.Stop(t)
+			testEnv.Clean(t)
```
Shouldn't we clean even if `d == nil` ?

dsheets replied:
This is how <https://github.com/moby/moby/blob/master/integration-cli/docker_cli_authz_unix_test.go#L77> and <https://github.com/moby/moby/blob/master/integration-cli/docker_cli_authz_plugin_v2_test.go#L43> are written. I've moved the cleanup out of the conditional in a subsequent commit.

dnephin commented on integration/plugin/authz/main_test.go:
```diff
@@ -0,0 +1,177 @@
+}
+
+func setupTest(t *testing.T) func() {
+	environment.ProtectImages(t, testEnv)
```
We merged a PR recently that added a ProtectAll() I think that should be used now. We should unexport the rest of these Protect*() method so it's more obvious.

dsheets replied:
If we used `ProtectAll`, wouldn't that negate the use of `testEnv.Clean`? I think that containers and plugins should probably be cleaned up after these tests, no?

dnephin replied:
`ProtectAll()` is maybe a misleading name. All it does is get a list of all the existing objects, so that `Clean()` doesn't remove them later. This is important for testing EE because it has system containers that it needs to keep running.

`Clean()` will still remove anything created by the tests.

dnephin commented on integration/plugin/authz/authz_plugin_v2_test.go:
```diff
@@ -0,0 +1,178 @@
+
+	// defer disabling the plugin
+	defer func() {
+		d.Restart(t)
```
If we are restarting the daemon on every test anyway why bother leaving it running and having to do extra cleanup?  Why not just stop it and recreate the daemon?

That should allow us to remove all the `d.Start()` at the end of test, which is really unexpected.

dsheets replied:
I believe the daemon is restarted in order to disable the authz plugin so that it does not interfere with cleanup. It may not be necessary but the original tests do this.

dnephin replied:
A big reason for creating this new test suite is to fix the numerous problems in the old suite. 

There are actually two daemons involved here. First there is a single daemon started before any of the test suites are run. Let's call this the "global daemon". Most tests hit this shared daemon, which is why Protect/Clean are required, so the next test can be run against a clean daemon.

These auth tests (also most swarm tests, or anything that requires specific daemon flags) start a new daemon within the test or test suite. Let's call this the "inner daemon".

There shouldn't be any reason to stop, restart, or clean up the inner daemon. We just stop it, and remove the root directory to wipe out any state. All these extra `Restart()` just add to confusion about the suite.

Could we try removing all of these `Restart()`calls and just do a proper clean by stopping and removing the temporary directory?

dnephin commented on integration/plugin/authz/authz_plugin_test.go:
```diff
@@ -0,0 +1,370 @@
+
+// assertURIRecorded verifies that the given URI was sent and recorded
+// in the authz plugin
+func assertURIRecorded(t *testing.T, uris []string, uri string) {
```
Could be replaced with `assert.Contains()` I think

dsheets replied:
Perhaps. It comes directly from <https://github.com/moby/moby/blob/master/integration-cli/docker_cli_authz_unix_test.go#L459>. It might be best to separate improving the test suite from porting it to not rely on the CLI.

dnephin replied:
If we start porting bad patterns from the old suite then we make it a lot more difficult to see the good patterns in the new suite. Can you try replacing it with `assert.Contains()`? It should be a small change.

dsheets replied:
It cannot be replaced with `assert.Contains` as it checks if the given `uri` is a substring of any of `uris` whereas `assert.Contains` checks for set membership using structural equality.

dnephin commented on integration/internal/main_test.go:
```diff
@@ -0,0 +1 @@
+package internal
```
`internal` shouldn't need a `main_test.go`

dnephin commented on integration/internal/api/volume/volume.go:
```diff
@@ -0,0 +1,61 @@
+	}
+
+	names := []string{}
+	for _, volume := range volumes.Volumes {
```
We should not be maintaining an abstraction over the APIClient for tests. Please remove this file. The only logic that really needs to be a new function is `namesFromVolumes()` which can be an unexported helper next to the one place it is used.

dsheets replied:
While it may be desirable in the long term to not have any abstraction over `APIClient` for tests, this design makes porting CLI tests much easier and reduces incidental complexity significantly. To maintain correctness, I recommend not attempting to simultaneously port tests and radically change them or instill new best practices.

Across existing tests, similar functionality will be required:

```
$ git grep "volume\", \"ls\")" integration-cli/ | wc -l
      13
```

```
$ git grep "volume\", \"ls" integration-cli/ | wc -l
      42
```

As this functionality is in `integration/internal`, golang's magic export scoping rules for `internal` will prevent users outside of this repository.

dnephin commented on integration/internal/api/system/system.go:
```diff
@@ -0,0 +1,45 @@
+}
+
+// Version provides the version of the daemon
+func Version(client client.APIClient) (types.Version, error) {
```
Please delete, this doesn't need to exist, just call `client.ServerVersion()` directly.

dsheets replied:
This exists to provide a homogeneous CLI replacement interface.

dnephin replied:
This is exactly the problem. We should not be testing as if we had a CLI. We need to test the API as directly as possible, not mimic CLI behaviour.

dsheets replied:
As explained in my PR comment, that is a laudable possible goal for the future but these tests were written against the CLI.

dnephin replied:
I guess I don't understand why we're moving these tests to `integration/` if they are going to be nearly identical to the `integration-cli` tests

dsheets replied:
My understanding is that we're *moving* the tests to eliminate the `moby/moby` ➡️ `docker/cli` dependency and unblock additional test contributions as `integration-cli/` is now deprecated.

dnephin commented on integration/internal/api/system/system.go:
```diff
@@ -0,0 +1,45 @@
+}
+
+// EventsSince returns event and error streams since a provided time
+func EventsSince(client client.APIClient, since string) (<-chan events.Message, <-chan error, func()) {
```
This can be an unexported helper next to the one place it is used.

dsheets replied:
It won't be used in one place for long. The `integration-cli` test suite has significant machinery for event stream handling.

dnephin replied:
We should only be testing events from the `integration/system` package, so it doesn't need to be exported.

dsheets replied:
A past authz test author clearly felt that it was necessary to test the interaction between events and authz plugins. In the future, it may be an improvement to move this test elsewhere but in the meantime it seems prudent test the integration between disparate components with the more complicated to test component. Once `integation-cli/` has been eliminated, it should be easy to review all the places where event functionality is used.

dnephin commented on integration/internal/api/plugin/plugin.go:
```diff
@@ -0,0 +1,59 @@
+
+// InstallGrantAllPermissions installs the plugin named and grants it
+// all permissions it may require
+func InstallGrantAllPermissions(client client.APIClient, name string) error {
```
This can be an unexported helper in the one file where it is used.

The rest of the file should be removed, they are one line wrappers around the client.

dsheets replied:
This will be used by other test suites -- at least mount point plugins. The other functions in this file exist to provide 1:1 mapping to the command line without depending on `docker/cli`. The client library imposes significant incidental complexity (contexts, options, questionable semantic decisions like in `PluginInstall`, etc) that a developer writing or porting a test shouldn't have to be subjected to. If, eventually, there is actually no benefit to fixing up the `APIClient` interface for test suite users, great. Until that time, this design makes it much easier to work with by e.g. not requiring questions like "Is there a utility function for this or is this one of the annoying `APIClient` functions?"

dnephin replied:
> 1:1 mapping to the command line

This is exactly what we do not want. If we were going to provide a 1:1 mapping to the CLI there would be little value in removing the CLI from the test suite.

If there are helpers that will be shared between the plugin suite then maybe create an `integration/plugin/internal` ?

dnephin reviewed (commented):
Thanks for porting this test suite. Mostly looks good, but I don't think `integration/internal/api/` should exist. Most of these functions are ` line wrappers around the api client which just hides the context. We should be testing the API more directly, not wrapping our client library with these helpers.

In a few cases there is some logic that can be extracted into factories for building the options struct, or helpers for transforming the responses, but I don't think they should wrap the client.

I think it's also better that we keep these as unexported. There should be very little re-use of these between test suites because each suite should be focused on just the component under test.

dnephin commented on integration/internal/api/image/image.go:
```diff
@@ -0,0 +1,66 @@
+}
+
+// Import imports the contents of a tarball named by path
+func Import(client client.APIClient, path string) error {
```
This can be replaced with a `ImageImportOptionsFromPath(path string) types.ImageImportOptions`

dsheets replied:
`git grep ImageImportOptionsFromPath` returns no results for me. It also doesn't look like it has a compatible signature. These functions are intended to be analogous to CLI invocations.

dnephin commented on integration/internal/api/image/image.go:
```diff
@@ -0,0 +1,66 @@
+)
+
+// Save saves an image to a tarball names by path
+func Save(client client.APIClient, path, image string) error {
```
Can be replaced with a `saveToPath(source io.Reader, path string) err`

dsheets replied:
It's not clear to me where `saveToPath` is implemented or how it can be a replacement with such a different signature.

dnephin commented on integration/internal/api/container/container.go:
```diff
@@ -0,0 +1,128 @@
+}
+
+// Export exports a container's file system as a tarball
+func Export(client client.APIClient, path, name string) error {
```
Looks like the same `saveToPath()` ?

dsheets replied:
It's not clear to me where `saveToPath` is implemented or how it can be a replacement with such a different signature.

dnephin replied:
I was thinking it would be implemented in `integration/plugin/authz` as an unexported helper. Sorry I didn't provide the full signature:

```go
savetoPath(t *testing.T, content io.ReadCloser) string
```
The return value is the path to the file.

dnephin commented on integration/internal/api/container/container.go:
```diff
@@ -0,0 +1,128 @@
+)
+
+// CreateVMemLabel creates a container with the supplied parameters
+func CreateVMemLabel(client client.APIClient, v []string, memoryMB int, labels map[string]string, img string, cmd []string) (string, error) {
```
I think all of these can be replaced with helpers that return `container.Config()` instead.

dsheets replied:
This would break the predictable homogeneity and make invocations of what used to be `docker run` much more verbose.

dnephin reviewed (commented):
Test helpers which accept `t *testing.T` as the first argument instead of returning `error` are also really nice. If we keep some of these helpers I would suggest passing in testing.T, which will help keep the tests readable and not distract from the relevant lines in the test.

dsheets commented:
> Thanks for porting this test suite. Mostly looks good, but I don't think integration/internal/api/ should exist. Most of these functions are ` line wrappers around the api client which just hides the context. We should be testing the API more directly, not wrapping our client library with these helpers.

The tests being ported are CLI tests. These wrappers homogenize the API client interface to make it behave more like the CLI interface and improve the porting experience.

> In a few cases there is some logic that can be extracted into factories for building the options struct, or helpers for transforming the responses, but I don't think they should wrap the client.

Why not? The tests do not demand most of the functionality that the API offers and which would clutter the tests with irrelevant noise.

> I think it's also better that we keep these as unexported. There should be very little re-use of these between test suites because each suite should be focused on just the component under test.

As the CLI -> API library is located in `internal`, it can't be used outside of the integration tests. My cursory grepping of the repository indicates that many CLI tests use a variety of commands that do not correspond directly to the specific functionality under test.

Do you have a document describing the `integration-cli` -> `integration` test migration plan?

Could you show me an `integration-cli` test suite you have ported that demonstrates how you'd like the port to be done and how to reduce the chance of defects during the port?

dnephin commented:
> The tests being ported are CLI tests. These wrappers homogenize the API client interface to make it behave more like the CLI

Exactly, that's the problem.

> The tests do not demand most of the functionality that the API offers and which would clutter the tests with irrelevant noise.

These are tests of the API. If you hide how the API is called the test loses a lot of value, because it's no longer clear what you are testing. The helpers may even silently break the test by masking the actual API behaviour. I agree that implementation details (like complex setup, and complex assertions) should be hidden from the testcase function, but the actual API call needs to be there, because that is what is being tested.

> many CLI tests use a variety of commands that do not correspond directly to the specific functionality under test.

Another major problem that will hopefully be fixed in the `integration/` suite.

> Do you have a document describing the integration-cli -> integration test migration plan?

There are a couple, but I will admit they are probably not as thorough as they could be. It is hard to enumerate all the problems with the old suite.

* https://github.com/moby/moby/issues/32866
* https://github.com/moby/moby/blob/master/TESTING.md
* https://forums.mobyproject.org/t/evolution-of-testing-moby/38

> Could you show me an integration-cli test suite you have ported that demonstrates how you'd like the port to be done and how to reduce the chance of defects during the port?

https://github.com/moby/moby/blob/master/integration/service/inspect_test.go there is also a test for version, but it is a very simple case.

Testing plugins seems to be more involved because it requires a daemon to be started with specific flags (which is not true of most of  the other API endpoints).

An important detail is that the API endpoint that is being tested is central to the test, and any setup/teardown is (as much as possible) extracted out into helpers with appropriately descriptive names. I think for the most part the tests here fit that criteria, the only different is the abstraction over the API calls.

If there are some common helpers that will be used between plugins, `integration/plugin/internal` might be a good place for them.

dnephin commented:
The plugin tests are also a very different case because they are actually testing API calls into other components. Some of these helpers for plugins are appropriate, but if we expose them to all of `integration/` they will be re-used incorrectly.

The ideal way to test these plugins would really be to re-use the test suite for the actual component. If we had the tests annotated correctly we could spin up a daemon that has the plugin loaded and then trigger the component test suite against that daemon. With the right test requirements/skips setup we wouldn't have to write a second test suite. We would have just a couple extra test cases for the "disallowed by plugin" case, which would get skipped when run against a non-plugin daemon.

I think that's probably unlikely to happen, so this suite is necessary.

dsheets commented:
It sounds like you have a lot of ideas for how `moby/moby` integration tests should work in the future. I opened this PR as it is all but required to satisfactorily test #33658. The goal here is reuse of existing test functionality to expand the authz test suite in a future-friendly way. I anticipate that this process of porting CLI tests to be independent of `docker/cli` will be required by other contributors who need to modify tests or contribute new tests. To that end, my goal is to make it as easy as possible to port CLI tests to be independent of `docker/cli`. This has several benefits:

1. `integration-cli` is drawn down and eliminated sooner as porting is easy
2. dependence on `docker/cli` is eliminated sooner
3. almost any contributor can port tests as the process is largely mechanical
4. incremental quality improvement is realized
5. the correspondence between CLI-esque and API use is immediately visible
6. any later SDK or client library author can see exactly and immediately the invocations necessary to achieve a CLI-like result
7. command executions are now named with symbols so they can be easily changed in the future

I think it's great that we have some ideas about how to improve the test suite in the future. In the present, having a mostly empty 'new' test suite and a large but deprecated 'old' test suite is negative for the project's health.

The existing tests are CLI tests (written as such) even if they *should* be testing the API or we *want* equivalent API tests. Actually testing the API directly will almost certainly look different from the current CLI tests and has its trade-offs. Couching CLI tests in the guise of API tests doesn't actually improve the tests but porting CLI tests to use the API under the hood does deliver the above benefits. Additionally, it seems to me that an ideal test suite would allow both 'low-level' API-style tests and 'high-level' CLI-style tests. As you note, some functionality like the plugins system greatly benefits from high-level tests.

It seems to me that any large scale improvement of the test suite should be done incrementally. Demanding that tests directly ported from `integration-cli` conform to some unwritten 'improved' specification seems counter-productive. The port itself (including functions that are 1:1 with the CLI) is valuable and speeds progress toward your long-term goals without demanding test rewrites by contributors or blocking other work. At a later stage, the CLI tests that have been ported can be rewritten or fixed up in uniform and well-defined stages as necessary.

dsheets commented:
> The ideal way to test these plugins would really be to re-use the test suite for the actual component. If we had the tests annotated correctly we could spin up a daemon that has the plugin loaded and then trigger the component test suite against that daemon. With the right test requirements/skips setup we wouldn't have to write a second test suite. We would have just a couple extra test cases for the "disallowed by plugin" case, which would get skipped when run against a non-plugin daemon.

I don't believe that design is feasible as these plugin tests target the plugin system's functionality that extends beyond a single plugin or accept/reject behavior. Examine the tests for daemon start failure or the instrumentation of authz v1 plugins in this PR for evidence.

dsheets commented:
> Test helpers which accept `t *testing.T` as the first argument instead of returning `error` are also really nice. If we keep some of these helpers I would suggest passing in testing.T, which will help keep the tests readable and not distract from the relevant lines in the test.

By not passing `t *testing.T` to the CLI analogs we retain the ability to test for expected failure.

Commit:
integration/plugin/authz: inline CLI work-alike library

I strongly disagree with the design of this pull request.

Signed-off-by: David Sheets <dsheets@docker.com>

dsheets commented:
I have rebased, squashed, and made the changes you requested. I strongly disagree with this design.

Please review and merge ASAP.

dnephin committed:
Skip all testdata in integration

Also skip.IfCondition directly from the test, so that the skip message is correct

Signed-off-by: Daniel Nephin <dnephin@docker.com>

dnephin reviewed (approved):
I pushed a small change to the integration path exclude list regex, and moved `skip.IfCondition()` into the test.

LGTM

dsheets commented:
So when can this be merged?

dnephin commented:
Needs another LGTM

vdemeester reviewed (approved):
LGTM 🐮
